### ISSUE TITLE
masterminds version incompatible with Drupal 8 core

### ISSUE THREAD

**Issue by ghost** [Lullabot/amp-library#110]
Drupal 8 core is locked to:

"name": "masterminds/html5",
"version": "2.1.2",

AMP requires:

"masterminds/html5": "^2.2.0",

They are incompatible :(

**Comment by sidkshatriya** [comment]
@karens 
I investigated this -- and this is a complicated issue. 

Firstly, why am I using `^2.2.0` of masterminds/html5 ... this is because of a bug in 2.1.2 (See https://github.com/Masterminds/html5-php/issues/98 ) which was fixed in 2.2.0

(See https://github.com/Masterminds/html5-php/releases for fixes per releaase)

Secondly, Drupal 8 core is not really incompatible with masterminds 2.2.0 and above.

If you look at `core/composer.json`

``` javascript
"require": {
        "php": ">=5.5.9",
...
        "masterminds/html5": "~2.1",
...    
}
```

"~2.1" means that the last digit can be changed e.g. 2.1, 2.2, 2.3 etc.

Now, Drupal core has chosen to lock its dependencies, sadly to a 1 year old version of the masterminds library.

(Here is a nice resource on `composer.lock` vs `composer.json` if you need a refresher https://blog.engineyard.com/2014/composer-its-all-about-the-lock-file )

So if I _did_ download a fresh installation of Drupal 8.1.x and typed:

``` bash
$ composer install 
$ composer require lullabot/amp
```

**I can confirm that I get a version constraint error**. This is of course because Drupal is locked down to the masterminds/html5 2.1.2 version while lullabot/amp wants `^2.2.0`.

Now I _wouldn't_ encounter this same issue if I had done the following

``` bash
$ composer install 
$ composer update
$ composer require lullabot/amp
```

Here the `composer update` updates masterminds to the latest version still satisfying the constraint `~2.1` which in this case is `2.2.1`. Now 2.2.1 is satisfies `^2.2.0` so the last composer require command succeeds.

#### Composer manager workflow

Now if you're using the composer manager workflow you should not encounter the same "incompatibility" issue.

composer manager asks you to download the drupal modules (e.g. amp) and then run `composer drupal-update` which is nothing but `composer drupal-rebuild && composer update` (See https://www.drupal.org/node/2405811 ). 

Notice the `composer update`

**So in this way this problem should not occur in drupal composer manager workflow**

#### Conclusion
- I think we should submit a change request to Drupal 8 maintainers to lock masterminds/html5 library at 2.2.1 instead which has many bug fixes. (Version 2.1.2 is a year old now)
- Workarounds to make our library work with _both_ 2.1.2 and 2.2.0 are complicated. When bug https://github.com/Masterminds/html5-php/issues/98 was found in the masterminds/html5 library, I added a workaround in https://github.com/Lullabot/amp-library/issues/22 (and all was good with version 2.1.2). However, the same workaround causes a regression in 2.2.0 ( https://github.com/Lullabot/amp-library/issues/53 ) so that workaround was removed and a dependency with masterminds `^2.2.0` added to the amp-library. However, now that has caused an issue with drupal's `composer.lock` using an outdated version as you have reported. But, as discussed, a `composer update` always solves this problem. We never hesitated in doing `composer update` before (e.g. in the drupal composer manager workflow) so I wonder if we should worry about this too much?
- Another approach would be to check if we're using 2.1.2 and then apply the workaround in #22, otherwise don't apply the workaround. This is likely to be hairy because I don't know of a smart way to get the version number of the mastermind/html5 library at runtime so I'll need to "test" the capability of the installation to see if the problem exists or not (sort of how browser capabilities are checked via javascript).

Any ideas?

**Comment by sidkshatriya** [comment]
@karens A interesting observation is that using a "pure" composer based distribution like https://github.com/acquia/lightning-project installs mastermind/html5 at `2.2.1` (!). So again you will not encounter the version incompatibility issue you are mentioning.

Try it out yourself:

``` bash
$ composer create-project acquia/lightning-project:^8.1.0 testlight --no-interaction
$ cd testlight
$ grep masterminds/html5 composer.lock -A10
```

(See https://github.com/acquia/lightning-project )
- The `composer.json` at https://github.com/acquia/lightning-project refers to `drupal/lightning` which is at https://github.com/acquia/lightning  
- The composer.lock at https://github.com/acquia/lightning is for mastermind 2.2.1 (!!) (See https://github.com/acquia/lightning/blob/8.x-1.x/composer.lock ). In other words the lightning project has chosen to update its composer.lock (for masterminds/html5 at least, probably for other packages too). So again you should not have problems using acquia lightning.

More and more, it seems to me that Drupal core needs to update their `composer.lock` file for masterminds/html5. 

This should not be not be a big deal. See https://github.com/drupal/drupal/commits/8.2.x/composer.lock for recent changes for other packages.

**Comment by sidkshatriya** [comment]
@karens Some more updates:
- There is a less heavy handed way to get this to work (in pure vanilla Drupal). Simply do:
  
  ``` bash
  $ composer update masterminds/html5
  ```
  
  instead of the (bazooka) approach of `$ composer update` which will update all possible dependencies.
- I filed a patch against Drupal core here: https://www.drupal.org/node/2752645 to update the composer.lock version for mastermind/html5. The good news is that updating the `composer.lock` file to masterminds/html5 version `2.2.1` **results in all tests passing with PHP 5.5 + mysql 5.5**. 

P.S. I've queued PHP 5.6 + mysql 5.5 and PHP 7.0 + mysql 5.5 and the results will be known in sometime.

**Comment by sidkshatriya** [comment]
@mtift FYI

**Comment by sidkshatriya** [comment]
**TL; DR**
- The issue occurs only if we insist on using the plain vanilla Drupal lock file. If you do `composer update masterminds/html5`(or simply `composer update`) then all is good. The lock file is likely to change anyways as you develop your project so I wonder whether this an important issue...
- A ticket+patch has been filed on drupal.org to update drupal core's `composer.lock` section for masterminds/html5 (See https://www.drupal.org/node/2752645 ). Version 2.1.2 is a year old!
- This issue does not occur on composer manager based installations or a pure composer based installation like acquia lightning

(To refresh yourself on composer.lock vs composer.json please read https://blog.engineyard.com/2014/composer-its-all-about-the-lock-file )

**Comment by ghost** [comment]
The particular problem I've run into is that I'm trying to add a new
requirement for another module (geofield). I tried to do composer require
geofield and then I tried to update composer.json and do composer update
geofield, and I get a response something like the following:

Loading composer repositories with package information

Updating dependencies (including require-dev)

Your requirements could not be resolved to an installable set of packages.

  Problem 1

```
- Conclusion: don't install masterminds/html5 2.2.1

- Conclusion: don't install masterminds/html5 2.2.0

- Installation request for masterminds/html5 (locked at 2.1.2, required
```

as ~2.1) -> satisfiable by masterminds/html5[2.1.2].

```
- lullabot/amp dev-master requires masterminds/html5 ^2.2.0 ->
```

satisfiable by masterminds/html5[2.2.0, 2.2.1, 2.x-dev].

```
- lullabot/amp dev-master requires masterminds/html5 ^2.2.0 ->
```

satisfiable by masterminds/html5[2.2.0, 2.2.1, 2.x-dev].

```
- lullabot/amp dev-master requires masterminds/html5 ^2.2.0 ->
```

satisfiable by masterminds/html5[2.2.0, 2.2.1, 2.x-dev].

```
- Conclusion: don't install masterminds/html5 2.x-dev
```

I'll keep poking at it, but that's where I've run into trouble.

**Comment by sidkshatriya** [comment]
Just do the following:

``` bash
$ composer update masterminds/html5
$ # do whatever else from here onwards. 
```

**Comment by sidkshatriya** [comment]
(typo: its `masterminds/html5` not `mastermind/html5`)

**Comment by ghost** [comment]
OK that worked. I wonder if other people will run into this and might need
instructions? But I'm back in business anyway.

**Comment by sidkshatriya** [comment]
@karens thats great!
- Long term fix: https://www.drupal.org/node/2752645 . This has the same workaround you used i.e. `composer update masterminds/html5` mentioned in the ticket. This workaround will be needed until Drupal core accepts the patch to their composer.lock file.
- People using composer manager (drupal 7 & 8) are not affected by this issue at all 
- People using acquia lightning not affected are not affected by this issue at all
- People that run `composer update` to refresh their dependencies (and their lock file) not affected by this issue.

Maybe I can add a short note somewhere on the amp module page for those people who still might be affected by this...

**Comment by sidkshatriya** [comment]
@mtift @karens Good news. The patch was accepted into Drupal core in double quick time for the 8.2.x and 8.1.x branches.

https://www.drupal.org/node/2752645#comment-11316123

**Comment by sidkshatriya** [comment]
@karens FYI I've added a short note to the README of 8.x

>  If you're not using Composer Manager, you _may_ run into a composer error in which you're unable to install the `masterminds/html5` library dependency and you get an error like "Your requirements could not be resolved to an installable set of packages." You may have run into https://www.drupal.org/node/2752645. See the link for a workaround. This problem will be fixed in Drupal 8.1.4

See http://cgit.drupalcode.org/amp/commit/?id=8775e8aec0cfd43c067e8994762b478d900560ab
